### DISCUSSION TITLE
Could ServiceLoaders be supported with Jbang?

### DISCUSSION THREAD

**Discussion by quintesse** [jbangdev/jbang#266] [General]
So working on a small Jbang project that could benefit from externally defined plugins I wondered how we could enable having an extensible system using ServiceLoaders. Right now it's not possible but there are already a couple of open issues that could bring us closer:

- ability to add resource files (to define services. See #39)
- the ability to include/depend on other jbang compiled source files (See #169)

To implement a system with a plugin I could imagine 4 files:

- spi.java
- app.java
- plugin.java
- META-INF/services/some.spi.interface

The `app.java` would have a `//DEPS spi.java` line that would cause the `spi.java` to be compiled and its JAR added to the class path (in case of `jbang run app.java`)

The `plugin.java` file would have both a `//DEPS spi.java` line and a `//FILES META-INF/services/some.spi.interface` line to add the file to the JAR's resources when built.

The `plugin.java` would never be `run` but only used with `build` (which can happen automatically so no need to run that manually).

And finally to run the application you'd do `jbang run --deps plugin.java app.java`

The `--deps` parameter would add the dependency to the plugin's JAR file at runtime where the application could look it up using the ServiceLoader system (compiling `plugin.java` as needed).

The nice thing is that this could work with existing features or ones that were already planned for inclusion.

Wdyt?

**Comment by maxandersen** [comment]
hmm - intriguing...so you want this:

![serviceloader](https://www.planttext.com/api/plantuml/svg/SoWkIImgAStDuUAABYx8rCjAB4kCvOfEB2Y0CGjoIjCpyw0y91a5NJi5X1O5AuMKr89YBbJrkDGW3D7J0wf0qMAbK9QPg92Lb9YNhg2h2z7Wa9gN0dGZ0000)

[source](https://www.planttext.com/?text=SoWkIImgAStDuUAABYx8rCjAB4kCvOfEB2Y0CGjoIjCpyw0y91a5NJi5X1O5AuMKr89YBbJrkDGW3D7J0wf0qMAbK9QPg92Lb9YNhg2h2z7Wa9gN0dGZ0000)

That would be kinda interesting to at least enable.

**Reply by quintesse** [reply]
Yuip, exactly that :-)

**Comment by dmlloyd** [comment]
Just came to :+1: this.  It would be very useful for plugin-driven systems to be able to give a `jbang` string in their "how to run this" documentation with a general way to make plugins visible for service loaders.

**Reply by maxandersen** [reply]
@dmlloyd just to be sure I grok your comment fully. The main thing this issue need is to allow `//DEPS` to refer to other jbang dependencies for then to be used with `--deps ` on command line to load "plugins".

Are you just agreeing that this is useful or are you thinking on something beyond that ?

**Reply by dmlloyd** [reply]
Just agreeing that it would be useful, especially for ad-hoc testing, one-off demos etc.
